### ISSUE TITLE
v4.1.0: History tab shows all versions for all Locales

### ISSUE THREAD

**Issue by chrispenny** [tractorcow/silverstripe-fluent#414]
## Locales

- International
-Default
-No fallback
- United States
-Fallback International
- Australia
-Fallback International
- Japan
-No fallback

### Created a new Page in `International`
<img width="774" alt="screen shot 2018-05-18 at 8 26 51 am" src="https://user-images.githubusercontent.com/505788/40202872-82902732-5a77-11e8-917b-fb0a9fee73e9.png">

### Localised Page in `Australia` with new metadata
<img width="771" alt="screen shot 2018-05-18 at 8 27 13 am" src="https://user-images.githubusercontent.com/505788/40202898-96ae1e9a-5a77-11e8-8fea-86b3e6fe3000.png">

### Localised Page in `United States` with new metadata
<img width="771" alt="screen shot 2018-05-18 at 8 27 51 am" src="https://user-images.githubusercontent.com/505788/40202932-a9601dfe-5a77-11e8-969a-3a3365c9f15d.png">

### Localised Page in `Japan` with new metadata
<img width="759" alt="screen shot 2018-05-18 at 8 31 37 am" src="https://user-images.githubusercontent.com/505788/40202943-b390d0c0-5a77-11e8-836d-804e0cf6a964.png">

## Result

Each Locale that you view the "History" tab in will show all versions for all Locales.

Each "History" tab will also have the most recent version selected - even if that versioned belongs to another Locale.

(`Japan` pictured below, but it's the same result in the other 3 Locales)

<img width="767" alt="screen shot 2018-05-18 at 8 32 07 am" src="https://user-images.githubusercontent.com/505788/40202966-c9318df2-5a77-11e8-80ba-35486501b07e.png">

**Comment by robbieaverill** [comment]
Relevant comment: https://github.com/tractorcow/silverstripe-fluent/issues/313#issuecomment-366563803

**Comment by chrispenny** [comment]
Ok, fair enough.

At the moment this is quite a confusing interface though - especially considering the top version is always selected, even when it's not the current version for the Locale you're in.

An indicator + the correct version (relative to your Locale) being selected when you load the page is probably reasonable. Bonus would be the ability to allow/disallow authors from reverting to a version from another Locale.

CC: @frankmullenger

**Comment by robbieaverill** [comment]
Yeah I agree. I think the logic in the CMS for generating the DataList of those versions is not built in a way that allows Fluent to modify it. Perhaps using the [history viewer](https://github.com/silverstripe/silverstripe-versioned-admin) with SilverStripe 4.2 onwards might help here

**Comment by chillu** [comment]
@robbieaverill Can you please classify this so it doesn't get lost? Are you using the `impact/*` labels to go through things on bug day? Ideally this gets looked at for your next bug day

**Comment by robbieaverill** [comment]
We need to discuss with @tractorcow when he's back, he's previously classed this as a feature

**Comment by tractorcow** [comment]
The long and short of it is:
 - a version number may be saved in a locale, but it doesn't belong to that locale per se; It's a little ambiguous at times, but every time you save in any locale, it does write to the base record, and that base content (e.g. the "test 3" in the above screenshot) is what's being shown to the user.
- We could just hide content saved in other locales, but there are cases where the user may need to see "all" versions despite. I would probably want a "show only this locale" option that you can turn on. However, that may be a pain to build.
- Filtering the history may be challenging, since you could unintentionally affect other internal API. Versioned::allVersions() does NOT let fluent extend it; It happens via extremely low level custom filters and table renames. CMSPageHistoryController itself has issues; The VersionsForm() method is extremely un-extendable. You couldn't easily do it without subclassing it and replacing it with DI.

tl;dr it's hard and may require core changes to fix.

**Comment by tractorcow** [comment]
Oh look, 5.0 uses Versioned::all_versions() instead, which IS extendable.

That means we can do it in 5.0 without core changes. :)

**Comment by chillu** [comment]
Is this only an issue in 4.1 which doesn't have the new GraphQL-based history viewer yet (due to be released in 4.2)? In this case, it's not worth fixing.

@tractorcow What module version are you referring to with "5.0" here?

**Comment by robbieaverill** [comment]
It might actually be susceptible to the same problem if it's using the same Versioned API. We've already found blocking problems with not being able to paginate or filter the versions that GraphQL gives you for versioned DataObjects (https://github.com/silverstripe/silverstripe-versioned/issues/147). We might need to backport the functionality in order to move forward with this kind of thing in 4.x

**Comment by tractorcow** [comment]
> @tractorcow What module version are you referring to with "5.0" here?

Recipe 5.x.

**Comment by tractorcow** [comment]
> It might actually be susceptible to the same problem if it's using the same Versioned API. We've already found blocking problems with not being able to paginate or filter the versions that GraphQL gives you for versioned DataObjects (silverstripe/silverstripe-versioned#147). We might need to backport the functionality in order to move forward with this kind of thing in 4.x

It seems that the limitation on that versioned API is affecting both fluent and graphql. Maybe we do need to back port it?

**Comment by tractorcow** [comment]
For recipe 4.2 we may be able to build a custom "lazy" list, a subclass of DataList, that returns the same Versioned_Version object. That would let us include the new functionality without breaking API / SemVer.

**Comment by robbieaverill** [comment]
Let's treat this as a bug I think

**Comment by robbieaverill** [comment]
So I've re-read this issue again after a few months, and basically the problem is that Versioned::allVersions() doesn't return a DataList, so Fluent can't modify the underlying queries.

SS5 has had a (breaking) rewrite and does return a DataList. I think the fix here is to backport that change in a non-breaking way then use that from `SilverStripe\Versioned\GraphQL\Operations\ReadVersions::resolve()`.

This would be only to allow Fluent to adjust the query for history viewer. From here we'd need to make sure that the returned list through Fluent is sensible

**Comment by robbieaverill** [comment]
Yep so with that change in place we still get all versions in all locales. We may need to do as @tractorcow suggested an add a filter to history viewer for "Only show versions from this locale" or something like that

**Comment by chrispenny** [comment]
Hi team (@robbieaverill),

I'm not sure if this is entirely related or not. Please let me know and I'll open up a separate issue if need be.

## Archive view shows the latest version (not Locale specific)

I have 2 Locales (neither fallback to anything):
- International
- Japan

I publish my Page in both Locales (International first, Japan second).

I then "Unpublish and archive" the Page in both Locales (International first, Japan second).

### Result

When you view the record, you get (I assume) a view of the most recent version. Using "Restore draft" will restore the version that you are viewing.

<img width="826" alt="screen shot 2018-09-25 at 7 31 46 am" src="https://user-images.githubusercontent.com/505788/45974315-1fbde980-c095-11e8-8f57-1bb1de192a67.png">

**Comment by robbieaverill** [comment]
@chrispenny I think it would be easier to track that issue separately - I've moved it to #471

**Comment by robbieaverill** [comment]
Weeee this is going to be tricky. It looks like Versioned adds an inner join to the versions table in order to get version information, but FluentVersionedExtension isn't rewriting that join.

What Fluent is doing at the moment is replacing SiteTree_Localised to SiteTree_Localised_Versions, which is fine, but when the table is SiteTree_Versions it's not being replaced with SiteTree_Localised_Versions.

I quickly tried making this replacement, but many of the columns required are in the base _Versions table, so we get errors like "WasDeleted" column not existing.

I think we need to put an inner join on any SiteTree_Versions tables to SiteTree_Localised_Versions, and fallback for any columns that don't exist to the SiteTree_Versions table. This is pretty tricky because the original inner join that we need to modify is added by Versioned as raw SQL, e.g. in `Versioned::augmentSQLVersionedLatest`:

https://github.com/silverstripe/silverstripe-versioned/blob/1/src/Versioned.php#L672-L689

Perhaps one option is to replace this with a SQLSelect implementation, which Fluent might automatically be able to rewrite

**Comment by robbieaverill** [comment]
>Perhaps one option is to replace this with a SQLSelect implementation, which Fluent might automatically be able to rewrite

It doesn't look like our DB layer allows this to happen. I think this might be a "can't fix" issue...

@tractorcow any suggestions?

**Comment by robbieaverill** [comment]
Hmm, the SQL query has a SourceLocale field added by `FluentExtension::augmentSQL` - if we can filter that by the current locale then it might work

**Comment by robbieaverill** [comment]
Alright so I've made some PRs that should hopefully resolve this by adding a filter to the DataList that the GraphQL resolver for version lists now returns (see the issue description for the PRs).

This now filters the versions in the history viewer by the correct locale that created them.

**Comment by chillu** [comment]
Apart from a fix for descending versions (https://github.com/silverstripe/silverstripe-cms/pull/2244) where the tests are still running, this has been merged.

**Comment by chrispenny** [comment]
Massive! Awesome effort. Thanks, Robbie (and everyone)!
